spotlight-client/src/VizHistoricalPopulationBreakdown/VizHistoricalPopulationBreakdownContainer.tsx

```diff
@@ -18,23 +18,55 @@
 import { observer } from "mobx-react-lite";
 import React from "react";
 import { withErrorBoundary } from "react-error-boundary";
+import { DataSeries } from "../charts/types";
 import HistoricalPopulationBreakdownMetric from "../contentModels/HistoricalPopulationBreakdownMetric";
+import {
+  DEMOGRAPHIC_UNKNOWN,
+  DIMENSION_DATA_KEYS,
+  DIMENSION_MAPPINGS,
+} from "../demographics";
 import ErrorMessage from "../ErrorMessage";
+import { HistoricalPopulationBreakdownRecord } from "../metricsApi";
+import { colors } from "../UiLibrary";
 import VizHistoricalPopulationBreakdown from "./VizHistoricalPopulationBreakdown";
 
-type VizHistoricalPopulationBreakdownContainerProps = {
+const VizHistoricalPopulationBreakdownContainer: React.FC<{
   metric: HistoricalPopulationBreakdownMetric;
-};
+}> = ({ metric }) => {
+  let chartData:
+    | DataSeries<HistoricalPopulationBreakdownRecord>[]
+    | null = null;
 
-const VizHistoricalPopulationBreakdownContainer: React.FC<VizHistoricalPopulationBreakdownContainerProps> = ({
-  metric,
-}) => {
-  if (metric.records)
-    return <VizHistoricalPopulationBreakdown data={metric.records} />;
+  const { demographicView, records } = metric;
 
-  if (metric.error) throw metric.error;
+  if (records) {
+    if (demographicView !== "nofilter") {
+      const labelsForDimension = DIMENSION_MAPPINGS.get(demographicView);
+      if (labelsForDimension) {
+        chartData = Array.from(labelsForDimension)
+          // don't need to include unknown in this chart;
+          // they are minimal to nonexistent in historical data and make the legend confusing
+          .filter(([value]) => value !== DEMOGRAPHIC_UNKNOWN)
+          .map(([value, label], index) => ({
+            label,
+            color: colors.dataViz[index],
+            coordinates:
+              demographicView === "total"
+                ? records
+                : records.filter(
+                    (record) =>
+                      record[DIMENSION_DATA_KEYS[demographicView]] === value
+                  ),
```

Comment:
This is the only place where I really question the location of the data transformation. My initial reaction is that the filtering should happen closer to the model and only the filtered data would ever reach this container (or even the chart itself), where the label and color mapping is applied.

Will the same transformation, both filtering and label/color mapping be applied for every chart that uses each particular Metric? If so, I wonder what it would look like to do all of this in the model and only pass the `metric.chartData` directly to the chart and skip this container altogether?

Reply:
good question ... there is kind of a two-stage filtering process here. The model filters for a specific "demographic lens" and then this code is splitting that data to create one series per demographic category (e.g., if the lens is "gender" there will be a series for "male" and for "female"). It's ... kind of specific to the chart library's API? But also kind of not, since the series are a structural property of the underlying data? Could go either way I guess, and since it was already here in the old component I just kind of left it here 😎 

I think maybe the container will also render the filter UI once I get to that, so I may still need it, but I think the idea of moving as much data transformation as possible into the model is good, so I will see if I can make that work.

Reply:
update: this worked and it feels massively better than doing it in the component. Good call!